identity/validator_set.go

```diff
@@ -366,6 +425,39 @@ func (vs *ValidatorStore) GetEndBlockUpdate(ctx *ValidatorContext, req types.Req
 				}
 			}
 		}
+
+		// purge all other active validators if not among the top
+		for addr, power := range vs.lastActive {
+			addrHuman := keys.Address(addr).Humanize()
+			pub, ok := noTopValidators[addrHuman]
+			if !ok {
+				continue
+			}
+			// get last purge height
+			purgeHeight, err := vs.GetLastPurgeHeight(keys.Address(addr))
+			if err != nil {
+				logger.Errorf("failed to get last purge height, validator: %s", addrHuman)
+				continue
+			}
+			// validator purged in block H persists in block H+1, H+2, so we can't purge it again
```

Comment:
So basically we don't need to purge 2 blocks after the original purge since the validator will persist in the list for those blocks. This means that it takes 2 blocks for the purge to take effect.

Reply:
> So basically we don't need to purge 2 blocks after the original purge since the validator will persist in the list for those blocks. This means that it takes 2 blocks for the purge to take effect.

Yes. The tendermint doc says that and it actually does that as well which crashed my node again in my first fix.